### DISCUSSION TITLE
.AsParallel() for IAsyncEnumerable

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/runtime#54583] [Q&A]
Is there something like this?  I'd like to fan my IAsyncEnumerable out to multiple processors just like I can with a regular enumerable.

**Comment by stephentoub** [comment]
```C#
await Parallel.ForEachAsync(yourAsyncEnumerable, async item =>
{
    ...
});
```

**Reply by jnm2** [reply]
I commonly have a similar need except the lambda returns a value and I need an ordered list of results at the end. I wish this was possible:

```cs
SomeResult[] results = await Parallel.ForEachAsync(
    asyncOrNonAsyncEnumerable,
    new ParallelOptions { MaxDegreeOfParallelism = 5 },
    async item =>
    {
        return await HitSomeApiAsync(item);
    });
```

I might start a proposal if I can't find anything like this.
